ml/nml.cc

```diff
@@ -0,0 +1,1133 @@
+        if (NumActiveDimensions)
+              host->host_anomaly_rate = static_cast<double>(host->mls.num_anomalous_dimensions) / NumActiveDimensions;
+
+        mls_copy = host->mls;
+
+        /*
+         * training stats
+        */
+        ts_copy = host->ts;
+
+        host->ts.queue_size = 0;
+        host->ts.num_popped_items = 0;
+
+        host->ts.allotted_ut = 0;
+        host->ts.consumed_ut = 0;
+        host->ts.remaining_ut = 0;
+
+        host->ts.training_result_ok = 0;
+        host->ts.training_result_invalid_query_time_range = 0;
+        host->ts.training_result_not_enough_collected_values = 0;
+        host->ts.training_result_null_acquired_dimension = 0;
+        host->ts.training_result_chart_under_replication = 0;
+
+        netdata_mutex_unlock(&host->mutex);
+    }
+
+    // Calc the avg values
+    if (ts_copy.num_popped_items) {
+        ts_copy.queue_size /= ts_copy.num_popped_items;
+        ts_copy.allotted_ut /= ts_copy.num_popped_items;
+        ts_copy.consumed_ut /= ts_copy.num_popped_items;
+        ts_copy.remaining_ut /= ts_copy.num_popped_items;
+
+        ts_copy.training_result_ok /= ts_copy.num_popped_items;
+        ts_copy.training_result_invalid_query_time_range /= ts_copy.num_popped_items;
+        ts_copy.training_result_not_enough_collected_values /= ts_copy.num_popped_items;
+        ts_copy.training_result_null_acquired_dimension /= ts_copy.num_popped_items;
+        ts_copy.training_result_chart_under_replication /= ts_copy.num_popped_items;
+    } else {
+        ts_copy.queue_size = 0;
+        ts_copy.allotted_ut = 0;
+        ts_copy.consumed_ut = 0;
+        ts_copy.remaining_ut = 0;
+    }
+
+    worker_is_busy(WORKER_JOB_DETECTION_DIM_CHART);
+    nml_update_dimensions_chart(host, mls_copy);
+
+    worker_is_busy(WORKER_JOB_DETECTION_HOST_CHART);
+    nml_update_host_and_detection_rate_charts(host, host->host_anomaly_rate * 10000.0);
+
+#ifdef NETDATA_ML_RESOURCE_CHARTS
+    worker_is_busy(WORKER_JOB_DETECTION_RESOURCES);
+    struct rusage PredictionRU;
+    getrusage(RUSAGE_THREAD, &PredictionRU);
+    updateResourceUsageCharts(RH, PredictionRU, TSCopy.TrainingRU);
+#endif
+
+    worker_is_busy(WORKER_JOB_DETECTION_STATS);
+    nml_update_training_statistics_chart(host, ts_copy);
+}
+
+typedef struct {
+    RRDDIM_ACQUIRED *acq_rd;
+    nml_dimension_t *dim;
+} nml_acquired_dimension_t;
+
+static nml_acquired_dimension_t nml_acquired_dimension_get(RRDHOST *rh, STRING *chart_id, STRING *dimension_id) {
+    RRDDIM_ACQUIRED *acq_rd = NULL;
+    nml_dimension_t *dim = NULL;
+
+    RRDSET *rs = rrdset_find(rh, string2str(chart_id));
+    if (rs) {
+        acq_rd = rrddim_find_and_acquire(rs, string2str(dimension_id));
+        if (acq_rd) {
+            RRDDIM *rd = rrddim_acquired_to_rrddim(acq_rd);
+            if (rd)
+                dim = reinterpret_cast<nml_dimension_t *>(rd->ml_dimension);
+        }
+    }
+
+    return nml_acquired_dimension_t{
```

Comment:
Codacy reports some issues here, so maybe using parentheses will fix it (I think as it is now it's valid C++ but not C syntax):
```
return (nml_acquired_dimension_t){
        .acq_rd = acq_rd,
        .dim = dim
    };
```

Reply:
Can you share the exact issue reported by Codacy?

Reply:
Vasili, I think he's referring to this -> https://app.codacy.com/gh/netdata/netdata/pullRequest?prid=11187669

Reply:
Allocated a temporary to make linter happy.

Reply:
Let me know if we are still getting the same issue again.

Reply:
That seems to have fixed it.